### ISSUE TITLE
Unable to load large number of tags on discover view

### ISSUE THREAD

**Issue by eddielopezfd** [getsentry/sentry#61999]
### Environment

SaaS (https://sentry.io/)

### Steps to Reproduce

- Visit any event here [https://fanduel-sb.sentry.io/discover/results/?field=title&field=release&field=environm[…]K-V2-EJF&sort=-timestamp&statsPeriod=90d&yAxis=count%28%29](https://fanduel-sb.sentry.io/discover/results/?field=title&field=release&field=environment&field=user.display&field=timestamp&name=EXC_BREAKPOINT%3A+Exception+6%2C+Code+13924416%2C+Subcode+8&project=5644496&query=issue%3ASPORTSBOOK-V2-EJF&sort=-timestamp&statsPeriod=90d&yAxis=count%28%29)  
- The page will error with a message on the bottom right corner "Tag cannot be loaded"

### Expected Result

Able to load page with tags

### Actual Result

some tags are unable to load on certain events within discover

### Product Area

Discover

### Link

https://fanduel-sb.sentry.io/discover/sportsbook-v2:c03cf6b3464e4b38a7abc002699ef073/?field=title&field=release&field=environment&field=user.display&field=timestamp&name=EXC_BREAKPOINT%3A+Exception+6%2C+Code+13924416%2C+Subcode+8&project=5644496&query=issue%3ASPORTSBOOK-V2-EJF&sort=-timestamp&statsPeriod=90d&yAxis=count%28%29

┆Issue is synchronized with this [Jira Improvement](https://getsentry.atlassian.net/browse/FEEDBACK-2049) by [Unito](https://www.unito.io)

**Comment by gggritso** [comment]
@eddielopezfd thanks for raising this! Is this a consistent issue, do you get the error every time, or has it gone away recently?

**Comment by gggritso** [comment]
@eddielopezfd hiya, thanks for raising this again. I'm going to take a look at this over the next few days and get back to you!

**Comment by eddielopezfd** [comment]
> @eddielopezfd hiya, thanks for raising this again. I'm going to take a look at this over the next few days and get back to you!

Thank you

**Comment by gggritso** [comment]
@eddielopezfd I took a look! I have some questions and some suggestions. First, the questions:

1. Where are you most hindered by this? Are you relying on tag autocomplete, and they're not showing up? Is the "Unable to load tags" message annoying? Something else?
2. How are you using your tags (I see you tag experiments and feature flags, which is common)? Do you need to be able to search by these tags, or are you only looking at them when looking at a specific exception?

**Comment by eddielopezfd** [comment]
> 1. Where are you most hindered by this? Are you relying on tag autocomplete, and they're not showing up? Is the "Unable to load tags" message annoying? Something else?

- Our team has noticed this issue primarily when viewing issues in Discover with a date range exceeding 24 hours. In these cases, valuable tags like route name or region fail to load, making it harder to analyze crashes and determine if they are isolated to specific areas or screens within the app.  The ability to add tags to larger date ranges (e.g., 7+ days) would also be beneficial for investigating long-standing crashes.

> 2. How are you using your tags (I see you tag experiments and feature flags, which is common)? Do you need to be able to search by these tags, or are you only looking at them when looking at a specific exception?

- While we occasionally filter by experiment or feature flag tags (especially after new releases or feature enablement), the primary concern is the inability to add essential tags like route name or region due to loading failures.

![Screenshot 2024-05-17 at 9 49 27 AM](https://github.com/getsentry/sentry/assets/97254677/f72f7394-429f-491c-83fb-75cea5b0fdd5)

**Comment by gggritso** [comment]
@eddielopezfd understood, thanks for sharing that!

First of all, I'm going to try to speed up the tag loading. Also, I think that maybe you should be able to type in whatever column you want, even if that tag wasn't loaded. I'll take that up with the product team.

Secondly, workarounds! You can force a tag through the URL. e.g., the URL `https://fanduel-sb.sentry.io/discover/homepage/?field=tags[region]&field=count()&query=&sort=-count&statsPeriod=24h&yAxis=count()` will create a table of issue counts by the `region` tag. This should work regardless of tag loading. You can substitute the tag you want in the right part of the URL to get the right query running. That'll at least unblock you when doing investigations.

Thirdly, an overall suggestion. Loading tags gets slower depending on how many events and unique tags a project has. If a project has a lot of events and a lot of unique tags (say, in the many 100s of tags), tag loading can get really slow. This degrades the product experience (tag loading, tag autocomplete, tags breakdown pages, etc.)

You could probably improve your overall experience with a few approaches:

1. Move any metadata that you _don't need to search on_ from a tag to a [context](https://docs.sentry.io/platforms/react-native/enriching-events/context/). A context can be a good place to put feature flags, experiments, or anything else that has a lot of data. You won't be able to search for this, but it'll show up on the Issue Details page, so it's handy while debugging. Contexts also have much looser restrictions on keys, values, and the overall size. They're a good fit for unstructured data.
2. If that's not possible, you can try being more restrictive about which tags are added. e.g., only add tags for the most interesting cases, only add a tag _if it's relevant_, only add a boolean tag if the value is true, etc.

**Comment by eddielopezfd** [comment]
Hi @gggritso , has there been any movement on getting this prioritized?

**Comment by gggritso** [comment]
@eddielopezfd not yet, sorry! I'll raise this with the team this week

**Comment by gggritso** [comment]
@eddielopezfd I'm working on this 👍🏻 I added a performance optimization here and I'll let you know when I enable it for your organization, should be early next week

**Comment by gggritso** [comment]
@eddielopezfd okay, update!

I'm working on some experimental changes to the tags autocomplete, and I enabled them for your organization. You should see fewer timeouts and faster response times for the tags (I'll be keeping an eye on the metrics). Let me know if your experience improves! The optimization applies for time ranges longer than two weeks.

One possible downside is that you _may_ find that in some UIs the tags autocomplete might not have all the tags you expect. This should not happen very often, but if it does let me know.

**Comment by gggritso** [comment]
@eddielopezfd I've been taking a look at the results of the change I made, and it looks like your organization isn't getting timeouts anymore. I'm going to close this, but I'm still tinkering with the optimizations.

If you see any problems with tags loading, or tag autocomplete, please file another issue! 🙏🏻
